### PR TITLE
Fixes #15206: Handling keept_output and asynchronous parameters in the remote-run API

### PR DESCRIPTION
**Pull request by ArmandSy**
https://issues.rudder.io/issues/15206

### CODE REVIEW

File: relay/sources/relayd/src/remote_run.rs [hunk 1]
```diff
@@ -65,6 +65,20 @@ impl RemoteRun {
         })
     }
 
+    pub fn reply_keep_output(&self, job_config: Arc<JobConfig>) -> impl warp::reply::Reply
+    {
+        if self.run_parameters.keep_output == true
```
Comment (amousset) [inline relay/sources/relayd/src/remote_run.rs]: no need to explicitely compare to `true`, `cargo clippy` would likely mention it

**Comment by ArmandSy** [comment]
Commit modified

**Comment by ArmandSy** [comment]
Commit modified

File: relay/sources/relayd/src/remote_run.rs [hunk 2]
```diff
@@ -65,32 +65,66 @@ impl RemoteRun {
         })
     }
 
+    pub fn target_to_vector(&self, job_config: Arc<JobConfig>) -> Vec<String> {
```
Comment (amousset) [inline relay/sources/relayd/src/remote_run.rs]: This should be a method of RemoteRunTarget, called `get_connected_nodes` or something like this.

Comment (amousset) [reply relay/sources/relayd/src/remote_run.rs]: Vector is a technical implementation detail, you should strive to use applicative business names as much as possible to keep things readable

File: relay/sources/relayd/src/remote_run.rs [hunk 3]
```diff
@@ -65,32 +65,66 @@ impl RemoteRun {
...
         &self,
         job_config: Arc<JobConfig>,
     ) -> Result<impl warp::reply::Reply, warp::reject::Rejection> {
+        if self.run_parameters.keep_output && self.run_parameters.asynchronous {
+            return Err(warp::reject::custom(
+                "keep_output and asynchronous cannot be true simultaneously",
+            ));
+        }
         if self.target == RemoteRunTarget::All {
             info!("conditions OK");
```
Comment (amousset) [inline relay/sources/relayd/src/remote_run.rs]: this is probably useless, if there is no error we can assume parameters were correct

**Comment by ArmandSy** [comment]
Commit modified

File: relay/sources/relayd/src/remote_run.rs (hunk 3 above)
Comment (amousset) [reply relay/sources/relayd/src/remote_run.rs]: up

**Comment by ArmandSy** [comment]
Commit modified

File: relay/sources/relayd/src/remote_run.rs [hunk 4]
```diff
@@ -221,8 +250,33 @@ impl RunParameters {
             .map_err(|e| panic!("error while running child: {}", e));
 
         tokio::spawn(child_future);
+        //thread::sleep(time::Duration::from_millis(500));
```
Comment (amousset) [inline relay/sources/relayd/src/remote_run.rs]: leftover from a test

File: relay/sources/relayd/tests/api_tests.rs [hunk 5]
```diff
@@ -47,4 +47,36 @@ mod tests {
 
         assert_eq!(res.unwrap().text().unwrap(), "Unhandled rejection: Invalid agent Condition : Invalid agent Condition : Wrong condition: \'clas~1\', it should match ^[a-zA-Z0-9][a-zA-Z0-9_]*$".to_string());
     }
+
+    #[test]
+    fn integration() {
```
Comment (amousset) [inline relay/sources/relayd/tests/api_tests.rs]: Could you give it a more descriptive name?

**Comment by ArmandSy** [comment]
Commit modified

**Comment by ArmandSy** [comment]
PR rebased

File: relay/sources/relayd/api_test.txt [hunk 6]
```diff
@@ -0,0 +1 @@
+remote run -D class2,class3 -H server.rudder.local
```
Comment (amousset) [inline relay/sources/relayd/api_test.txt]: This is the result of the test, it should not be included.

Comment (amousset) [reply relay/sources/relayd/api_test.txt]: Furthermore, temporary files created by tests should go into `target/tmp`

File: relay/sources/relayd/tests/api_test.txt [hunk 7]
```diff
@@ -0,0 +1 @@
+hello moto la
```
Comment (amousset) [inline relay/sources/relayd/tests/api_test.txt]: same here

**Comment by ArmandSy** [comment]
Commit modified

**Comment by ArmandSy** [comment]
Commit modified

File: relay/sources/relayd/tests/relay_api.rs [hunk 8]
```diff
@@ -47,4 +47,36 @@ mod tests {
...
+            ("classes", "class2,class3"),
+            ("nodes", "server.rudder.local"),
+        ];
+
+        client
+            .post("http://localhost:3030/rudder/relay-api/remote-run/nodes")
+            .form(&params)
+            .send();
+
+        let data = std::fs::read_to_string("./api_test.txt").expect("Unable to read file");
```
Comment (amousset) [inline relay/sources/relayd/tests/relay_api.rs]: This path needs to be updated

**Comment by ArmandSy** [comment]
Commit modified

**Comment by ArmandSy** [comment]
Commit modified

**Comment by ArmandSy** [comment]
Commit modified

**Pull request merged**